src/thor/timedep_forward.cc

```diff
@@ -136,57 +145,114 @@ bool TimeDepForward::ExpandForward(GraphReader& graphreader,
     // Decide if we should expand a shortcut or the non-shortcut edge...
 
     // We didn't add any shortcut of the uturn, therefore evaluate the regular uturn instead
-    disable_uturn = ExpandForwardInner(graphreader, pred, nodeinfo, pred_idx, uturn_meta, tile,
-                                       offset_time, destination, best_path) ||
-                    disable_uturn;
+    disable_uturn =
+        ExpandInner<expansion_direction>(graphreader, pred, opp_pred_edge, nodeinfo, pred_idx,
+                                         uturn_meta, tile, offset_time, destination, best_path) ||
+        disable_uturn;
   }
 
   return disable_uturn;
 }
 
+template bool TimeDepForward::Expand<TimeDepForward::ExpansionType::reverse, BDEdgeLabel>(
+    GraphReader& graphreader,
+    const GraphId& node,
+    BDEdgeLabel& pred,
+    const uint32_t pred_idx,
+    const DirectedEdge* opp_pred_edge,
+    const TimeInfo& time_info,
+    const valhalla::Location& destination,
+    std::pair<int32_t, float>& best_path);
+
 // Runs in the inner loop of `ExpandForward`, essentially evaluating if
 // the edge described in `meta` should be placed on the stack
 // as well as doing just that.
 //
 // Returns true if any edge _could_ have been expanded after restrictions etc.
-inline bool TimeDepForward::ExpandForwardInner(GraphReader& graphreader,
-                                               const EdgeLabel& pred,
-                                               const NodeInfo* nodeinfo,
-                                               const uint32_t pred_idx,
-                                               const EdgeMetadata& meta,
-                                               const graph_tile_ptr& tile,
-                                               const TimeInfo& time_info,
-                                               const valhalla::Location& destination,
-                                               std::pair<int32_t, float>& best_path) {
+template <TimeDepForward::ExpansionType expansion_direction, typename EdgeLabelT>
+inline bool TimeDepForward::ExpandInner(GraphReader& graphreader,
+                                        const EdgeLabelT& pred,
+                                        const baldr::DirectedEdge* opp_pred_edge,
+                                        const NodeInfo* nodeinfo,
+                                        const uint32_t pred_idx,
+                                        const EdgeMetadata& meta,
+                                        const graph_tile_ptr& tile,
+                                        const TimeInfo& time_info,
+                                        const valhalla::Location& destination,
+                                        std::pair<int32_t, float>& best_path) {
+  constexpr bool FORWARD = expansion_direction == TimeDepForward::ExpansionType::forward;
+
+  if (!FORWARD) {
```

Comment:
Question to readers: conceptually why is this only done for the reverse search?  Is this a copypasta bug, or is there some rationale?

cc @kevinkreiser @dnesbitt61

Reply:
Code has changed considerably since I left it, but in reverse searches I initially added an early exit based on the directed edge's reverseaccess to avoid having to get the opposing directed edge and perform the more complete costing Allowed check. I believe I did some benchmarking around this and found a slight performance improvement. I can't comment on why shortcut edges are skipped.